### PR TITLE
Comparator searches

### PR DESCRIPTION
**Pull request by czh-orz**
Adding tests to search by comparator. Also made made a more generic way of resolving items by path for a given resource. can_resolve_element can receive a codeblock and only return true if anything in the given can make the codeblock return true.

### CODE REVIEW

File: generators/uscore-r4/generator.rb [hunk 1]
```diff
@@ -306,6 +300,44 @@ def get_search_params(search_parameters, sequence)
...
+    when 'Period', 'date'
+      comparators.each do |comparator, expecatation|
+        next unless ['SHALL', 'SHOULD'].include? expecatation
+
+        comparator_val_name = "#{comparator}_#{param.tr('-', '_') + '_val'}"
+        case comparator.to_s
+        when 'gt', 'ge'
+          search_code += %(\n
+        #{comparator_val_name} = '#{comparator}' + (DateTime.xmlschema(#{param_val_name}) - 1).xmlschema)
+        when 'lt', 'le'
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply generators/uscore-r4/generator.rb]: One is +1 and the other is -1

File: generators/uscore-r4/generator.rb [hunk 2]
```diff
@@ -322,69 +354,77 @@ def search_param_constants(search_parameters, sequence)
...
     case type
-    when 'CodeableConcept'
+    when 'Period'
       search_validators += %(
-        when '#{element}'
-          codings = #{path_parts.join('&.')}#{'&.first' if contains_multiple}&.coding
-          assert !codings.nil?, '#{element} on resource did not match #{element} requested'
-          assert codings.any? { |coding| !coding.try(:code).nil? && coding.try(:code) == value }, '#{element} on resource did not match #{element} requested'
+          comparator = value[0, 1]
+          value = value[2..-1] if ['ge', 'gt', 'le', 'lt'].include? comparator
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply generators/uscore-r4/generator.rb]: value[0,1] returns a string if value is a string. But it should actually by value[0,2].

File: generators/uscore-r4/generator.rb [hunk 3]
```diff
@@ -322,69 +354,77 @@ def search_param_constants(search_parameters, sequence)
...
+            when 'ge'
+              end_date >= value_date || end_date.nil?
+            when 'le'
+              start_date <= value_date || start_date.nil?
+            when 'gt'
+              end_date > value_date || end_date.nil?
+            when 'lt'
+              start_date < value_date || start_date.nil?
+            else
+              valueDate >= start_date && valueDate <= end_date
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply generators/uscore-r4/generator.rb]: For now, the only comparators that are shoulds or shalls are these four.

File: generators/uscore-r4/generator.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jammjammjamm) [reply generators/uscore-r4/generator.rb]: Ahh, ok. I think `[]` with a range is a lot more readable than `[]` with two arguments, so I'd either change the first line to use `[0..1]`  or `take(2)` and `drop(2)`.

File: generators/uscore-r4/generator.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jammjammjamm) [reply generators/uscore-r4/generator.rb]: If we're only supporting those four, then I think the `else` should raise an error. As it is now, the `else` code is incorrect but a newcomer reading this wouldn't have any reason to believe that it is any less correct than the code around it.

File: generators/uscore-r4/generator.rb [hunk 4]
```diff
@@ -306,6 +300,44 @@ def get_search_params(search_parameters, sequence)
...
+    search_assignments << "'#{param}': #{param_value_name(param)}"
+  end
+  search_assignments_str = "{ #{search_assignments.join(', ')} }"
+  search_params.each do |param|
+    param_val_name = param_value_name(param)
+    param_info = sequence[:search_param_descriptions][param.to_sym]
+    comparators = param_info[:comparators]
+    type = param_info[:type]
+    case type
+    when 'Period', 'date'
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply generators/uscore-r4/generator.rb]: I decided to do a case in case we need to add other types in the future. Right now only period and date types are required by US Core.

File: generators/uscore-r4/generator.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (czh-orz) [reply generators/uscore-r4/generator.rb]: one is +1 and the other is -1.

File: generators/uscore-r4/generator.rb [hunk 5]
```diff
@@ -322,69 +354,77 @@ def search_param_constants(search_parameters, sequence)
...
       # https://www.hl7.org/fhir/search.html#string
-      search_validators +=
-        if contains_multiple
-          %(
-        when '#{element}'
-          found = #{path_parts.join('&.')}&.any? do |name|
-            name.text&.include?(value) ||
+      search_validators += %(
+          value_found = can_resolve_path(resource, '#{path_parts.join('.')}') do |name|
+            name.text&.include(value) ||
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply generators/uscore-r4/generator.rb]: I've updated it to use start_with for everything except family and text and normalized the capitalization. This (https://www.hl7.org/fhir/search.html) says to normalize the accents and combined characters of other languages though, which I'm not sure what to do with.

File: generators/uscore-r4/generator.rb [hunk 6]
```diff
@@ -322,69 +354,77 @@ def search_param_constants(search_parameters, sequence)
...
               name.family.include?(value) ||
               name.given.any { |given| given&.include?(value) } ||
-              name.prefix.any { |prefix| prefix.include?(value) } ||
-              name.suffix.any { |suffix| suffix.include?(value) }
+              name.prefix.any { |prefix| prefix&.include?(value) } ||
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply generators/uscore-r4/generator.rb]: I think so.... FHIR says "The multiple given parts and family name combine to form a single name." Do you think it should only match the first?

File: generators/uscore-r4/generator.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jammjammjamm) [reply generators/uscore-r4/generator.rb]: I would definitely vote against adding useless code to the repo because it might become useful later. It is just as possible that when we need to add more types we determine that a different approach is necessary and then this code never becomes useful.

[the remark this answers is not in the document]
Comment (czh-orz) [reply generators/uscore-r4/generator.rb]: I don't know. Using case matches more closely with the appropriate underlying logic to me. I mean the alternative is to just do a 'next unless' which doesn't really match what I had in mind for the function. Like yeah, it's technically not necessary to do a case, but I think it fits better aesthetically, and it's only adding one line anyways. Anyways, if you feel strongly about it, I'm willing to just go with whatever.

File: generators/uscore-r4/generator.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (Jammjammjamm) [reply generators/uscore-r4/generator.rb]: No, I was just commenting on the method name `.any` vs `.any?`.

File: generators/uscore-r4/generator.rb [hunk 7]
```diff
@@ -231,45 +233,43 @@ def create_references_resolved_test(sequence)
...
-    path_parts[index] = 'local_class' if part == 'class' # match fhir_models because class is protected keyword in ruby
-    parts_with_multiple << index if element_descriptions[cur_path.downcase.to_sym][:contains_multiple]
-  end
-  parts_with_multiple.each do |index|
-    path_parts[index] += '&.first'
-  end
   path_parts[0] = "@#{path_parts[0].downcase}"
-
-  path_parts.join('&.')
+  resource_val = path_parts.shift
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply generators/uscore-r4/generator.rb]: Fixed

File: generators/uscore-r4/generator.rb [hunk 8]
```diff
@@ -306,6 +300,44 @@ def get_search_params(search_parameters, sequence)
   search_code
 end
 
+def get_comparator_searches(search_params, sequence)
+  search_code = ''
+  search_assignments = []
+  search_params.each do |param|
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply generators/uscore-r4/generator.rb]: Fixed

File: generators/uscore-r4/generator.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (czh-orz) [reply generators/uscore-r4/generator.rb]: I changed to used range.

File: generators/uscore-r4/generator.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (czh-orz) [reply generators/uscore-r4/generator.rb]: Fixed.

File: generators/uscore-r4/generator.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (czh-orz) [reply generators/uscore-r4/generator.rb]: I added the other comparators

File: generators/uscore-r4/metadata_extractor.rb [hunk 9]
```diff
@@ -139,7 +139,7 @@ def add_must_support_elements(profile_definition, sequence)
           sequence[:must_supports] <<
             {
               type: 'element',
-              path: path.gsub('[x]', type['code'])
+              path: path.gsub('[x]', type['code'].slice(0).capitalize + type['code'].slice(1..-1))
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply generators/uscore-r4/metadata_extractor.rb]: Yeah, for things like effective + dateTime, it should be effectiveDateTime

File: lib/app/sequence_base.rb [hunk 10]
```diff
@@ -741,14 +741,52 @@ def check_resource_against_profile(resource, resource_type, specified_profile =
...
 
-        el_as_array.any? { |el| can_resolve_path(el.send(cur_path_part), path_ary.join('.')) }
+        if block_given?
+          el_as_array.any? { |el| can_resolve_path(el.send(cur_path_part), path_ary.join('.')) { |value_found| yield(value_found) } }
+        else
+          el_as_array.any? { |el| can_resolve_path(el.send(cur_path_part), path_ary.join('.')) }
+        end
+      end
+
+      def resolve_element_from_path(element, path)
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply lib/app/sequence_base.rb]: I moved them to another file and included it into sequencebase

File: generators/uscore-r4/generator.rb [hunk 11]
```diff
@@ -322,69 +344,99 @@ def search_param_constants(search_parameters, sequence)
...
+            when 'lt'
+              start_date < value_date || start_date.nil?
+            when 'ne'
+              value_date < start_date || value_date > end_date
+            when 'sa'
+              start_date > value_date
+            when 'eb'
+              end_date < value_date
+            when 'ap'
+              true # don't have a good way to check this
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply generators/uscore-r4/generator.rb]: I ended up implementing it as search range overlaps with target range since that was the wording in the FHIR search page (though it also says something about 10%). Anyways, this should never be hit  anyways since we're not searching by 'ap' during testing.

File: generators/uscore-r4/generator.rb [hunk 12]
```diff
@@ -322,69 +344,99 @@ def search_param_constants(search_parameters, sequence)
...
+            when 'ge'
+              end_date >= value_date || end_date.nil?
+            when 'le'
+              start_date <= value_date || start_date.nil?
+            when 'gt'
+              end_date > value_date || end_date.nil?
+            when 'lt'
+              start_date < value_date || start_date.nil?
+            when 'ne'
+              value_date < start_date || value_date > end_date
```
[the remark this answers is not in the document]
Comment (czh-orz) [reply generators/uscore-r4/generator.rb]: I ended up changing the implementation completely and it should match the FHIR spec now.

File: lib/app/utils/search_validation.rb [hunk 13]
```diff
@@ -0,0 +1,139 @@
+# frozen_string_literal: true
+
+module Inferno
+  module SearchValidationUtil
+    def can_resolve_path(element, path)
```
Comment (Jammjammjamm) [inline lib/app/utils/search_validation.rb]: I think we should leave this and `resolve_element_from_path` in `SequenceBase` for now since they're general methods which are used in contexts other than search date validation. I suspect that these might make the most sense as extensions on FHIR models so that we can do something like `resource.resolve? some_path`, but that is outside the scope of this PR.

File: generators/uscore-r4/generator.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (Jammjammjamm) [reply generators/uscore-r4/generator.rb]: It looks like this change disappeared at some point.

File: generators/uscore-r4/generator.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (arscan) [reply generators/uscore-r4/generator.rb]: I think it's a good idea for us to avoid placeholder code as well.  Also, if another type with a comparator suddenly shows up and the code in here isn't generic enough to handle it, then we should raise an exception.  So could you get rid of the `case` and put in the following guard clause (or something like it), if it makes sense to you?

```
raise 'Only Period and date types for search comparators supported' unless ['Period','date'].include? type
